Тетрадка предназначена для студентов образвательного проекта Booster MDS использование вне проекта запрещено.

##### Задания не несут смысловой нагрузки для реальной жизни, они сделаны лишь для того, чтоб научить использовать Pandas

#### Каждое задание оценивает в 1 балл, для успешной сдачи задания нужно набрать 12 баллов

This database contains 76 attributes, but all published experiments refer to using a subset of 14 of them. In particular, the Cleveland database is the only one that has been used by ML researchers to this date. The "goal" field refers to the presence of heart disease in the patient. It is integer valued from 0 (no presence) to 1.

https://www.kaggle.com/ronitf/heart-disease-uci

___

In [2]:
import pandas as pd
import numpy as np

In [3]:
data = pd.read_csv('d:\heart_pandas.csv')

In [4]:
data.head(5)

age     sex   chest_pain_type  resting_blood_pressure  cholesterol  \
0   63    male  non-anginal pain                     145          233   
1   37    male   atypical angina                     130          250   
2   41  female    typical angina                     130          204   
3   56    male    typical angina                     120          236   
4   57  female    typical angina                     120          354   

     fasting_blood_sugar               rest_ecg  max_heart_rate_achieved  \
0  greater than 120mg/ml                 normal                      150   
1    lower than 120mg/ml  ST-T wave abnormality                      187   
2    lower than 120mg/ml                 normal                      172   
3    lower than 120mg/ml  ST-T wave abnormality                      178   
4    lower than 120mg/ml  ST-T wave abnormality                      163   

  exercise_induced_angina  st_depression   st_slope  num_major_vessels  \
0                      no            2.3  upsloping                  0   
1                      no            3.5  upsloping                  0   
2                      no            1.4       flat                  0   
3                      no            0.8       flat                  0   
4                     yes            0.6       flat                  0   

    thalassemia  target  
0        normal       1  
1  fixed defect       1  
2  fixed defect       1  
3  fixed defect       1  
4  fixed defect       1

#### Описание признаков

**age**: The person's age in years

**sex**: The person's sex (1 = male, 0 = female)

**cp**: The chest pain experienced (Value 1: typical angina, Value 2: atypical angina, Value 3: non-anginal pain, Value 4: asymptomatic)

**trestbps**: The person's resting blood pressure (mm Hg on admission to the hospital)

**chol**: The person's cholesterol measurement in mg/dl

**fbs**: The person's fasting blood sugar (> 120 mg/dl, 1 = true; 0 = false)

**restecg**: Resting electrocardiographic measurement (0 = normal, 1 = having ST-T wave abnormality, 2 = showing probable or definite left ventricular hypertrophy by Estes' criteria)

**thalach**: The person's maximum heart rate achieved

**exang**: Exercise induced angina (1 = yes; 0 = no)

**oldpeak**: ST depression induced by exercise relative to rest ('ST' relates to positions on the ECG plot. See more here)

**slope**: the slope of the peak exercise ST segment (Value 1: upsloping, Value 2: flat, Value 3: downsloping)

**ca**: The number of major vessels (0-3)

**thal**: A blood disorder called thalassemia (3 = normal; 6 = fixed defect; 7 = reversable defect)

**target**: Heart disease (0 = no, 1 = yes)

> 1. age 
> 2. sex 
> 3. chest pain type (4 values) 
> 4. resting blood pressure 
> 5. serum cholestoral in mg/dl 
> 6. fasting blood sugar > 120 mg/dl
> 7. resting electrocardiographic results (values 0,1,2)
> 8. maximum heart rate achieved 
> 9. exercise induced angina 
> 10. oldpeak = ST depression induced by exercise relative to rest 
> 11. the slope of the peak exercise ST segment 
> 12. number of major vessels (0-3) colored by flourosopy 
> 13. thal: 3 = normal; 6 = fixed defect; 7 = reversable defect

___

**Сколько мужчин в датасете? Сколько женщен? (sex) **

In [80]:
res = data['sex'].value_counts()

In [85]:
print('мужчин -',res[0], '; женщин -', res[1])

мужчин - 207 ; женщин - 96


 **Какой процент мужчин в датасете? (решите в одну строчку, не используя предыдущий результат. Не забудте знак процента) **

In [6]:
print(data['sex'].value_counts('male')['male']*100,'%')

68.3168316832 %


Ответ: мужчин в датасете - 68.32%

**Сколько мужчин имеют заболевание сердца? Сколько женщин имеют заболевание сердца? **

In [75]:
men = data[data['sex']=='male'][data['target'] == 1]
women = data[data['sex']=='female'][data['target'] == 1]
print('мужчин с заболеванием сердца: ', men['sex'].shape[0], 'женщин с заболеванием сердца: ', women['sex'].shape[0])

мужчин с заболеванием сердца:  93 женщин с заболеванием сердца:  72


d:\soft\miniconda\lib\site-packages\ipykernel_launcher.py:1: UserWarning: Boolean Series key will be reindexed to match DataFrame index.
  """Entry point for launching an IPython kernel.
d:\soft\miniconda\lib\site-packages\ipykernel_launcher.py:2: UserWarning: Boolean Series key will be reindexed to match DataFrame index.
  


Ответ: мужчин - 93, женщин - 72

**Какую долю, от общего числа пациентов, занимают мужчины не имеющие болезнь сердца?**

In [48]:
data[(data['sex'] == 'male') & (data['target'] == 0)].shape[0] / data.shape[0]

0.37623762376237624

Ответ: 0.38 от всей выборки

**Сколько лет самому молодому пациенту, страдающему болезнью сердца?**

In [33]:
data[data['target'] == 1]['age'].min()

29

Ответ: 29 лет

**Сколько лет самому возастному пациенту, у которого нет проблем с сердцем?**

In [34]:
data[data['target'] == 0]['age'].max()

77

Ответ: 77 лет

**Сколько лет самой молодой женщине, которая страдает болезнью сердца?**

In [66]:
print('возраст самой молодй женщины: ',data[(data['target'] == 1) & (data['sex'] == 'female')]['age'].min())

возраст самой молодй женщины:  34


Ответ: 34 года

**Какой средний возраст женщин?**

In [67]:
print('средний возраст женщин: ', data[data['sex'] == 'female']['age'].mean())

средний возраст женщин:  55.6770833333


Ответ: средний возраст всех женщин из датасета - 56 лет

**Каковы средние значения и среднеквадратичные отклонения возраста тех, кто страдают болезнью сердца?**

In [68]:
print('среднее значение возраста: ', data[data['target'] == 1]['age'].mean(), 
      '\nсреднеквадратичное отклонение возраста: ', data[data['target'] == 1]['age'].std())

среднее значение возраста:  52.496969697 
среднеквадратичное отклонение возраста:  9.55065075195


Ответ: среднее значение возраста:  52.5 года, среднеквадратичное отклонение возраста:  9.55 лет

** Правда ли, что люди не болеющие болезнью сердца имеют уровня холестерина меньше среднего? (chol) **

In [69]:
avg_chol = data['cholesterol'].mean()
avg_chol_for_no_heart_deseases = data[data['target'] == 0]['cholesterol'].mean()
if(avg_chol_for_no_heart_deseases < avg_chol): 
    print("Yes")
else:
    print("No")

No


Ответ: нет, не правда

**Выведите статистику rest_ecg для все числовых признаков, его максимально и среднее значение (используйте groupby(), решите в одну строчку)**

In [35]:
## Ваш ответ:

**Посчитайте у кого уровень депрессии при физический нагрузке выше (в среднем), среди мужчин страдающих болезнью сердца или среди женщин не страдающих болезнью сердца (st_depression)**

In [88]:
male_lvl_of_depression = data[(data['sex'] == 'male') & (data['target'] == 1)]['st_depression'].mean()
female_lvl_of_depression = data[(data['sex'] == 'female') & (data['target'] == 0)]['st_depression'].mean()

In [89]:
if(male_lvl_of_depression > female_lvl_of_depression):
    print('Уровень депрессии выше среди мужчин с болезнями сердца')
else:
    print('Уровень депрессии выше среди женщин не страдающих болезнями сердца')

Уровень депрессии выше среди женщин не страдающих болезнями сердца


**Посчтитайте максимальный и минимальный уровень холестерина для каждого типа chest_pain_type, rest_ecg, thalassemia. Пишите код оптимально, можно использовать циклы**

In [ ]:
## Ваш код

In [ ]:
## Ваш ответ:

**Сколько значений может принимать каждый из категориальных признаков?**

In [96]:
res = data.describe(include=['object'])
print(res)

         sex chest_pain_type  fasting_blood_sugar               rest_ecg  \
count    303             303                  303                    303   
unique     2               3                    2                      3   
top     male  typical angina  lower than 120mg/ml  ST-T wave abnormality   
freq     207             193                  258                    152   

       exercise_induced_angina   st_slope   thalassemia  
count                      303        303           303  
unique                       2          2             3  
top                         no  upsloping  fixed defect  
freq                       204        161           166  


**У какого категориального признака наблюдается самый сильный дизбаланс классов?**

In [40]:
## Ваш код

In [ ]:
## Ваш ответ:

---

@Booster
Мамкин Data Scientist

tg @mommyscience